### DISCUSSION TITLE
OAuth all-in-one popup blocked on Safari

### DISCUSSION THREAD

**Discussion by j0sh** [pocketbase/pocketbase#3563] [Q&A] [answered]
Hi - really liking Pocketbase so far. I have an issue with the OAuth all-on-one popup being blocked on MacOS Safari.

As far as I can tell, all pop-ups are blocked by default on MacOS Safari. They are still blocked even after user interaction, eg click to login.

Popups can be allowed on a case-by-case basis, but the indicator in the address bar is subtle and easy to miss, so it's not a great UX to depend on.

Is this a known issue and/or is there a workaround for using the OAuth all-in-one flow with Safari? The manual code exchange flow is certainly an option, but all-in-one would be nice to have if possible.

This is not using iOS or a PWA, so it seems different than https://github.com/pocketbase/pocketbase/discussions/2429 - the solutions there did not help.

In case it helps, here is my [login page](https://gist.github.com/j0sh/9e0aab279107f9eac664ef307ba9f0bd) - pretty vanilla React + Typescript setup (with Vite).

In https://github.com/pocketbase/pocketbase/discussions/2906#discussioncomment-6468116 there was this comment:

> The "all-in-one" flow was added primarily to help with the integration in environments where redirect/deep-link handling is not always possible or easy to handle (eg. Figma/AdobeXD plugins, SPA apps, CLI apps, mobile apps, etc.)

Does that mean purely browser-based environments are *not* a good fit for the all-in-one flow? If so, it might be good to more explicitly state this in the [documentation](https://pocketbase.io/docs/authentication/), which seems to recommend the all-in-one flow without caveats.

**Comment by ganigeorgiev** [comment, marked as answer]
I don't have Safari or MacOS to test it.

What version of the JS SDK do you use?

Which solution from #2429 did you try? Did you try the one with the eager `window.open` from https://github.com/pocketbase/pocketbase/discussions/2429#discussioncomment-5943061? The latest JS SDK do this also automatically but in your case since you are wrapping it in your own async function, you may want to try open the window **before** the async declaration at the beginning of `callGithub` function (in your case you may also have to wrap it in another regular function because it seems that it will be executed before the onClick has taken place).

**Reply by ganigeorgiev** [reply]
To rule out React/env issue and to simplify the test, here is static index.html that you can put in a `pb_public` directory next to the executable:

```js
<!-- pb_public/index.html -->
<!DOCTYPE html />
<html>
    <head>
        <meta charset="utf-8" />
        <title>Test auth</title>
    </head>
    <body>
        <button id="btn">Login</button>

        <script type="module">
            import PocketBase from "https://cdn.jsdelivr.net/gh/pocketbase/js-sdk@0.19.0/dist/pocketbase.es.mjs"

            const pb = new PocketBase("http://127.0.0.1:8090");

            document.getElementById("btn").addEventListener("click", (e) => {
                e.preventDefault();

                pb.collection("users").authWithOAuth2({
                    provider: "github"
                }).then((result) => {
                    console.log("Success:", result);
                }).catch((err) => {
                    console.log("Error:", err)
                });

                // if it doesn't work, try to explicitly eager open the window
                // let w = window.open() 
                // pb.collection("users").authWithOAuth2({
                //     provider: "github",
                //     urlCallback: (url) => {
                //         w.location.href = url;
                //     },
                // }).then((result) => {
                //     console.log("Success:", result);
                // }).catch((err) => {
                //     console.log("Error:", err)
                // });
            })
        </script>
    </body>
</html>
```

Open  `http://127.0.0.1:8090` in your browser and click on the "Login" button. The above is expected to work without triggering the browser popup blocker.

**Reply by j0sh** [reply]
Thanks for the quick reply.

The JS SDK version is 0.19.0 (latest on yarn)

> Which solution from https://github.com/pocketbase/pocketbase/discussions/2429 did you try? Did you try the one with the eager window.open from https://github.com/pocketbase/pocketbase/discussions/2429#discussioncomment-5943061?

Yes, I tried that one - both the eager `w = window.open()` open before entering the async function, and assigning a plain `window.location.href = url`

> in your case you may also have to wrap it in another regular function because it seems that it will be executed before the onClick has taken place

Amazingly, I think this is what did the trick. So it actually works now! 🎉  Thank you very much for the help.

In summary, something like this is required for _async_ click handlers on Safari:

```js
function clickHandler(ev:React.MouseEvent<HTMLAnchorElement>){
  ev.stopPropagation();
  const w = window.open()
  if (!w) return false;
  (async function(){
    await  pb.collection("users").authWithOAuth2({
      provider: "github",
      urlCallback: (url) => {
        w.location.href = url
      }
    });
  })()
  return false;
}
```

If async/await is not being used, this seems to also work:
```js
function clickHandler(e){
   e.preventDefault();
  pb.collection("users").authWithOAuth2({
    provider: "github",
    urlCallback: (url) => {
      window.location.href = url;
    }
  }).then((result) => {
    console.log("Success:", result);
  }).catch((err) => {
    console.log("Error:", err)
  });
}
```

I am not sure why one would require `window.open()` while the other works with `window.location` alone, but hey it works 🤷
